tests/api/batch/test_utils.py

```diff
@@ -0,0 +1,176 @@
+    is_failure_logged = BatchTileStatus.FAILED in tile_status_sequence[-1]
+    assert logging_mock.call_count == int(is_processing_logged) + int(is_failure_logged)
+
+
+def _tile_status_counts_to_tiles(tile_status_counts: Dict[BatchTileStatus, int]) -> List[Dict[str, str]]:
+    """From the info about how many tiles should have certain status it generates a list of tile payloads with these
+    statuses.
+
+    Each payload should be approximately what Sentinel Hub service returns but because we don't need all parameters we
+    just return status. At the end we randomly shuffle the list just to make it more general.
+    """
+    tiles: List[Dict[str, str]] = []
+    for tile_status, count in tile_status_counts.items():
+        for _ in range(count):
+            tiles.append({"status": tile_status.value})
+
+    random.shuffle(tiles)
+    return tiles
+
+
+def test_monitor_batch_job_sleep_time_error() -> None:
+    with pytest.raises(ValueError):
+        monitor_batch_job("x", sleep_time=59)
+
+    with pytest.raises(ValueError):
+        monitor_batch_job("x", analysis_sleep_time=4)
+
+
+@pytest.mark.skipif(sys.version < "3.8", reason="Mocking check for call.args doesn't work correctly for Python 3.7")
+@pytest.mark.parametrize(
+    "status_sequence",
+    [
+        (BatchRequestStatus.CREATED, BatchRequestStatus.ANALYSING, BatchRequestStatus.ANALYSIS_DONE),
+        (BatchRequestStatus.DONE,),
+        (BatchRequestStatus.ANALYSING, BatchRequestStatus.CANCELED),
+        (BatchRequestStatus.FAILED,),
+    ],
+)
+@pytest.mark.parametrize("config", [SHConfig(), None])
+@pytest.mark.parametrize("sleep_time", [5, 1000])
+def test_monitor_batch_analysis(
+    status_sequence: Tuple[BatchRequestStatus, ...], config: SHConfig, sleep_time: int, mocker: MockerFixture
+) -> None:
+    """This test mocks:
+
+    - a method of Batch API interface to avoid calling Sentinel Hub service,
+    - sleeping time to avoid waiting,
+    - logging to ensure logs are being recorded.
+
+    At the end it also checks if all mocks have been called the expected number of times and with expected parameters.
+    """
+    batch_requests = [
+        BatchRequest(request_id="mocked-request", process_request={}, tile_count=42, status=status)
+        for status in status_sequence
+    ]
+
+    batch_mock = mocker.patch("sentinelhub.SentinelHubBatch.get_request")
+    batch_mock.side_effect = batch_requests
+    sleep_mock = mocker.patch("time.sleep")
+    logging_mock = mocker.patch("logging.Logger.info")
+
+    if status_sequence[-1] in [BatchRequestStatus.CANCELED, BatchRequestStatus.FAILED]:
+        with pytest.raises(RuntimeError):
+            monitor_batch_analysis("mocked-request", config=config, sleep_time=sleep_time)
+    else:
+        result = monitor_batch_analysis("mocked-request", config=config, sleep_time=sleep_time)
+        assert result is batch_requests[-1]
+
+    sleep_loop_counts = len(status_sequence) - 1
```

Comment:
as above, i'm not sure we should be checking the implementation details :thinking: especially since so many other methods are called in the background

Reply:
I believe we should in this case. We can discuss more about this in person.